### ISSUE TITLE
Broken link?

### ISSUE THREAD

**Issue by krytarowski** [hgeisse/netbsd-on-eco32#1]
I cannot open this page:

curl: (7) Failed to connect to vshg02.mni.thm.de port 80: Connection refused

**Comment by hgeisse** [comment]
Yes, the link in the README isn't valid any longer. I replaced it with a note saying that the porting project's homepage is down, and that it will take quite a while before I'm able to bring it up again. Thanks for the hint.

**Comment by krytarowski** [comment]
Any chances to get it functional with NetBSD and merge with mainline? We will need toolchain and simulator.

**Comment by hgeisse** [comment]
Not with NetBSD, sorry, I have no time to work on that. But you can of course get the current simulator version (this is my project "hgeisse/eco32" here on GitHub). Besides that, Martin Hofmann (mh0fmann here on GitHub) has just recently completed a Linux port to ECO32, with a suitable GCC toolchain (written by Jens Mehler some time ago). He plans to publish the whole thing on GitHub within the next few weeks. If you need further help, please feel free to contact me or Martin by email.

**Comment by krytarowski** [comment]
I will keep using GitHub as more convenient.

I was looking at eco32 repository and it looks impressive with all the toolchain programs, but I'm afraid that they won't be suitable for building NetBSD.

How have you been building NetBSD? With this patched gnu gcc and gnu binutils? How to obtain them? 

How to run simulator with eco32 and partial NetBSD support?

**Comment by hgeisse** [comment]
Hi Kamil,

On Do, 2019-02-14 at 10:40 -0800, Kamil Rytarowski wrote:
> I will keep using GitHub as more convenient.

that's fine with me.

> I was looking at eco32 repository and it looks impressive with all the
> toolchain programs, but I'm afraid that they won't be suitable for
> building NetBSD.

Thanks. The toolchain contained therein was suficient to bring up a
port of UNIX 7th Edition to ECO32. I'm teaching courses on operating
systems at our university and I needed a small system that the students
could experiment with. But of course this toolchain isn't suited for
building NetBSD.

> How have you been building NetBSD? With this patched gnu gcc and
> gnu binutils?

NetBSD was our first attempt at porting a contemporary OS to ECO32.
Jens Mehler and I implemented the necessary BFD support for the
GCC toolchain, and Jens constructed an ECO32 back-end for gcc.
We both then brought NetBSD to the point of booting - not quite
stable, but eventually we got a shell prompt. Jens then left
academia for a job in industry, and I didn't have the time to
pursue the project any further.

A year or so later, a few good students showed interest in porting
an OS to ECO32, but they were far more interested in Linux than in
NetBSD. So they started all over again, this time with Linux as the
target. And as I already told you, we had success :-)

> How to obtain them?

I will take a look into the safety copies of our repositories;
the then current version should be recoverable. Give me a few
days to explore that. But be warned: NetBSD has been developing
further, and the GCC toolchain underwent big changes. Do not
expect to get something easily runnable.

This, by the way, is IMO the biggest obstacle for maintaining
a port. You need at least two or three really competent people,
who are willing to devote quite some time for this rather
unrewarding task.

> How to run simulator with eco32 and partial NetBSD support?

I ported NetBSD's mkfs to be run under Linux and writing to a
partition of the simulated disk. Together with the ECO32 machine
monitor, which allows booting a selected partition, it was
possible to get the NetBSD image into memory. I will gladly
explain the gory details if you are interested, but must first
recover the old repository contents.

Best regards,
Hellwig

**Comment by krytarowski** [comment]
> I will gladly explain the gory details if you are interested, but must first recover the old repository contents.

Yes please. Please show how to use it (or how it was used) possibly with command lines step by step.

**Comment by hgeisse** [comment]
With a bit of work, I was able to recover the old repository. You can fetch it from here:
http://vshg02.mni.thm.de/~bouncer/NetBSD/NetBSD-on-ECO32.tar.gz
When you unpack the archive, you will find:
  - NetBSD 6.1.5 source packages
  - patches for ECO32 against 6.1.5
  - a short HOWTO

There is a build script within the patches, which used to build a runnable NetBSD (tools,
kernel, userland - sort of). I tried that, but the build crashes on an up-to-date Ubuntu.
I think that this is nothing serious, but I really don't have the time to look into the problem.
As Chaucer put it: "The lyf so short, the craft so longe to lerne."  :-)

Good luck!
Hellwig

**Comment by krytarowski** [comment]
Thanks! I'm going to play with it!

**Comment by OhSoKuul** [comment]
Hi krytarowski, I was one of the students who helped porting NetBSD to eco32 back in the day. I found my backup of my thesis about it - complete with all sources and how to build it. So far I was able to build the toolchain and the kernel. The userland throws some errors but those shouldn't be a deal-breaker I guess. 

The main difference is that hgeisse uses 6.1.X and my verson is still 6.0 - I am currently experimenting with both builds - I will keep you posted. If you have any questions you can contact me :)

I was able to build tools and kernel for eco32 as explained in the HOWTO file. I have no pre-setup on my system it was basically a one-time shot to see if that stuff still works. I recommend sticking to hgeisse's version of netbsd-on-eco32 - it's still deprecated netbsd but not as deprecated as my backup

**Comment by krytarowski** [comment]
Resurrecting the usable userland part will help. Thank you for update.

**Comment by OhSoKuul** [comment]
I will try to see what I can do with my limited time and maybe I can get the userland part working again.

**Comment by OhSoKuul** [comment]
Okay I got the userland build running on my system again

Apply the patch to usr/src/eco32.sh to make it more portable.


I have to admit that I am using macOS 10.14.X with LLVM and some homebrew stuff.
I couldn’t get it to build under a recent ubuntu 18.X because the toolchain around gcc and binutils
are way to new to build netbsd 6.X in all it’s glory.

If you want to build under ubuntu you have to use a GCC version 5 or lower I guess ( I don’t have the time to figure that out on my - sorry)

But generally speaking the project still builds fine :)

The patch:

--- eco32.sh.orig	2019-02-19 18:09:52.000000000 +0100
+++ eco32.sh	2019-02-19 18:13:16.000000000 +0100
@@ -73,7 +73,7 @@
 # instructions to build more libs after return 0 - this can be done by hand if needed
 function build_cross_compiler_libs
 {
-	ECO32_MAKE=$(locate nbmake-eco32)
+	ECO32_MAKE=$(find . -type f -name "nbmake-eco32" | xargs -I {} echo "$PWD/"{})
 	if [ $? -eq 1 ] 
 		then
 			echo "Searching for nbmake-eco32 via 'sudo updatedb'"

**Comment by Okarin88** [comment]
After that I ran into some more problems:
patch:
--- eco32.sh.orig	2019-02-19 18:09:52.000000000 +0100
+++ eco32.sh	2019-02-19 21:26:15.000000000 +0100
@@ -9,10 +9,10 @@
 export MKSTATICLIB=yes
 
 # setup the NetBSD build-environment - we don't need, want or use this stuff -- disableing it makes build go faster
-# MKVARS="MKMAKEMANDB MKDYNAMICROOT MKLLVM MKGROFF MKGCCCMDS MKCXX MKGCC MKBINUTILS MKATF MKBSDTAR MKCATPAGES MKCOMPAT MKCOMPLEX MKCRYPTO MKCRYPTO_RC5 MKCVS MKDEBUG MKDEBUGLIB MKDOC MKDYNAMICROOT MKEXTSRC MKGDB MKHESIOD MKHTML MKIEEEFP MKINET6 MKINFO MKIPFILTER MKISCSI MKLDAP MKLINT MKLVM MKMAN MKMANDOC MKMDNS MKNLS MKNPF MKPAM MKPCC MKPF MKPIC MKPICINSTALL MKPICLIB MKPOSTFIX MKPROFILE MKSHARE MKSKEY MKX11FONTS MKX11 MKYP USE_HESIOD USE_INET6 USE_KERBEROS USE_LDAP USE_PAM USE_SKEY USE_SSP USE_FORT USE_YP"
+MKVARS="MKMAKEMANDB MKDYNAMICROOT MKLLVM MKGROFF MKGCCCMDS MKCXX MKGCC MKBINUTILS MKATF MKBSDTAR MKCATPAGES MKCOMPAT MKCOMPLEX MKCRYPTO MKCRYPTO_RC5 MKCVS MKDEBUG MKDEBUGLIB MKDOC MKDYNAMICROOT MKEXTSRC MKGDB MKHESIOD MKHTML MKIEEEFP MKINET6 MKINFO MKIPFILTER MKISCSI MKLDAP MKLINT MKLVM MKMAN MKMANDOC MKMDNS MKNLS MKNPF MKPAM MKPCC MKPF MKPIC MKPICINSTALL MKPICLIB MKPOSTFIX MKPROFILE MKSHARE MKSKEY MKX11FONTS MKX11 MKYP USE_HESIOD USE_INET6 USE_KERBEROS USE_LDAP USE_PAM USE_SKEY USE_SSP USE_FORT USE_YP"
 # old line needs to be kept for documentation
 # share/mk/bsd.README has the explanation for this variables
-MKVARS="MKLINT MKDYNAMICROOT MKLLVM MKGROFF MKGCCCMDS MKCXX MKGCC MKBINUTILS MKATF MKEXTSRC MKGDB MKPIC MKPICINSTALL MKPICLIB MKPROFILE MKX11FONTS MKX11"
+#MKVARS="MKCRYPTO MKLINT MKDYNAMICROOT MKLLVM MKGROFF MKGCCCMDS MKCXX MKGCC MKBINUTILS MKATF MKEXTSRC MKGDB MKPIC MKPICINSTALL MKPICLIB MKPROFILE MKX11FONTS MKX11"
 
 
 function userland_help
@@ -73,7 +73,7 @@
 # instructions to build more libs after return 0 - this can be done by hand if needed
 function build_cross_compiler_libs
 {
-	ECO32_MAKE=$(locate nbmake-eco32)
+	ECO32_MAKE=$(find . -type f -name "nbmake-eco32" | xargs -I {} echo "$PWD/"{})
 	if [ $? -eq 1 ] 
 		then
 			echo "Searching for nbmake-eco32 via 'sudo updatedb'"
@@ -94,7 +94,7 @@
 			
 			echo "preparing for userland build"
 			ECO32_DEST=$(cat $ECO32_MAKE | grep "DESTDIR=" | sed "s/DESTDIR='//g" | sed "s/'; export DESTDIR//g")
-			
+			HDEFINES="-I$PWD/tools/gcc/obj/build/gcc/ -I$PWD/include/ -I$PWD/obj/destdir.eco32/usr/include/"
 			pushd .
 				cd tools/gcc/
 				$ECO32_MAKE HAVE_GCC=45 bootstrap-libgcc || exit 1 
@@ -129,17 +129,17 @@
 				cd tools/gcc/obj/build
 				cp Makefile Makefile.bk
 				#teach the gcc the new includes pathes :)
-				sed -i 's/^FLAGS_FOR_TARGET = /FLAGS_FOR_TARGET = -B$(tooldir)\/usr\/lib -isystem $(tooldir)\/usr\/include/g' Makefile
+				sed -i 's/^FLAGS_FOR_TARGET = /FLAGS_FOR_TARGET = -B$(tooldir)\/usr\/lib -I $(tooldir)\/usr\/include/g' Makefile
 				sed -i "s/^FLAGS_FOR_TARGET = /tooldir=$(echo $ECO32_DEST | sed 's/\//\\\//g')\nFLAGS_FOR_TARGET = /" Makefile
 				make configure-target-libgcc || exit 6
+                                sed -i "s|INCLUDES\ =|INCLUDES\ =\ $HDEFINES\ |g" eco32--netbsd/libgcc/Makefile
 				make all-target-libgcc || exit 6
 				make install-target-libgcc || exit 6
 
 				#cd tools/gcc/obj/build
-
-				make configure-target-libiberty || exit 7
-				make all-target-libiberty || exit 7
-				make install-target-libiberty || exit 7
+				#make configure-target-libiberty || exit 7
+				#make all-target-libiberty || exit 7
+				#make install-target-libiberty || exit 7
 
 				#defect 
 				#make configure-target-libstdc++-v3 || exit 8
@@ -150,9 +150,9 @@
 				#make all-target-libgomp || exit 9
 				#make install-target-libgomp || exit 9
 
-				make configure-target-libobjc || exit 10
-				make all-target-libobjc || exit 10
-				make install-target-libobjc || exit 10
+				#make configure-target-libobjc || exit 10
+				#make all-target-libobjc || exit 10
+				#make install-target-libobjc || exit 10
 			
 				cp Makefile.bk Makefile
 			popd

**Comment by OhSoKuul** [comment]
The main problem is that I can't get the build-env from 2014 back :D Too much has changed. Right now I even doubt that we used the netbsd userland ... maybe we went straigt to busybox or something like that :D 

I modyfied the eco32.sh to just build libgcc and make it swallow the header files but that can't be the solution ... this was all so long ago
